File: 0416-partition-equal-subset-sum/0416-partition-equal-subset-sum.cpp
class Solution {
public:
    bool func(vector<int> & nums, vector<vector<int>> &dp,int target,int index){
        
        if(target==0){
            return true;
        }
        if(index<0){
            return  false;
        }

        if(index==0){
            if(target==nums[0]){
                return true;
            }
        }
        if(dp[index][target]!=-1){
            return dp[index][target];
        }

        bool not_take = func(nums,dp,target,index-1);

        // take or not take 
        bool take = false;
        if(nums[index]<=target){
            take = func(nums,dp,target-nums[index],index-1);
        }

        
        return dp[index][target] = take || not_take;


    }




    bool canPartition(vector<int>& nums) {
        // convert into subsets 
        int n = nums.size();
        long long int sum =0;
        for(auto k : nums){
            sum += k;
        }

        if(sum%2!=0){
            return false;
        }
        int target = sum/2;

        vector<vector<int>> dp(n,vector<int>(target+1,-1));

        return func(nums,dp,target,n-1);


    }
};

File: 0416-partition-equal-subset-sum/0416-partition-equal-subset-sum.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        # partition equal subset sum 
        # return such that the we can have equal sum 

        sum = 0
        n = len(nums)
        
        for x in nums:
            sum += x
        
        if sum%2!=0:
            return False
        
        target = sum//2

        def func(ind,target):

            # base case 
            if target ==0:
                return True
            
            if ind ==0:
                if target%nums[ind]==0:
                    return True
                else:
                    return False

            # take and not take 
            take = 0
            if nums[ind] <=target:
                take = func(ind-1,target-nums[ind])
            
            # not take
            not_take = func(ind-1,target)


            return take or not_take


        return func(n-1,target)

        